website/blog/2023-10-24-sqlsmith-hero.md

```diff
@@ -0,0 +1,104 @@
+- **Performance Tests**: Validate performance impact of new features and optimizations. Prevent performance regressions.
+
+While these tests ensure functional correctness and stability during rapid development, they have limitations. Test SQL queries are mostly handwritten and simple, lacking complex queries seen in real-world scenarios. They have insufficient coverage of edge cases and exceptions. 
+
+## Introducing SQLsmith
+
+[SQLsmith](https://github.com/anse1/sqlsmith) is a randomized SQL query generator that produces a high volume of diverse test cases to simulate real-world variability. Compared to other testing approaches, SQLsmith improves test coverage to uncover more potential issues and bugs.
+
+## Implementing Our SQLsmith
+
+The original SQLsmith for PostgreSQL fuzz testing was inspired by [Csmith](https://github.com/csmith-project/csmith). Now, many known open source databases have their own SQLsmith ports like [CockroachDB](https://github.com/cockroachdb/cockroach/blob/master/pkg/workload/sqlsmith/sqlsmith.go), [TiDB](https://github.com/PingCAP-QE/go-sqlsmith), [RisingWave](https://github.com/risingwavelabs/risingwave/tree/main/src/tests/sqlsmith) etc. 
+
+These open source SQLsmiths use diverse languages (C++, Go, Rust) with different syntax support suitable for their domains. We couldn't directly use them and had to build our own SQLsmith in Rust to fully support Databend's syntax and features.
+
+SQLsmith has three main components:
+
+- **SQL Generator**: Generates abstract syntax trees (ASTs) of various types
+- **SQL Reducer**: Simplifies failing SQLs to isolate bug causes
+- **Runner**: Executes SQLs, records errors 
+
+## SQL Generator
+
+The SQL Generator randomly generates ASTs including:
+
+- **Data Types**: Basic types, Nested types like `Array`, `Map`, `Tuple` etc.
+- **DDL**: `CREATE`, `ALTER`, `DROP TABLE` etc. to create and modify test tables.
+- **DML**: `INSERT`, `UPDATE`, `DELETE`, `MERGE` etc. to populate test data.
+- **Query**: `WITH`, `SELECT`, `JOIN`, `SubQuery`, `ORDER BY` etc. `WITH` and `SubQuery` can generate complex queries.
+- **Expression**: `Column`, `Literal`, `Scalar Function` and `Aggregate Function` etc. Expressions can be nested.
+
+By randomly generating AST components, it covers all possible SQL syntax. Controlled recursion also creates complex nested SQLs to find obscure bugs. But depth is capped to avoid unexecutable SQLs.
+
+## SQL Reducer
+
+Since generated SQLs can be very complex with the bug trigger buried, simplifying the original SQL makes bug isolation easier.
+
+The SQL Reducer iteratively removes AST components like `WITH`, `SubQuery`, `Expression` etc. If the simplified SQL still reproduces the bug, it's used. Else original SQL is retained. Ultimately, it outputs the smallest SQL to reproduce the bug by traversing all AST parts.
+
+## Runner 
+
+The SQL Runner loops generating test SQLs, executes them using Databend, and processes failures as:
```

Comment:
```suggestion
The SQL Runner executes test SQL queries using Databend, and in case of failure, it proceeds with the following error-handling steps:
```